### PR TITLE
Sovereign cloud support for AzureGermanCloud and AzureUSGovCloud.

### LINKED ISSUE #1066: Sovereign cloud support for AzureGermanCloud and AzureUSGovCloud
Add sovereign cloud support for AzureGermanCloud and AzureUSGovCloud

### PR DESCRIPTION
**Pull request by wangtt03**
Changes:
* Extract all the sovereign cloud configurations to `defaults.go`
* Cleanup hard coded URL suffix in shell scripts.
* Remove FQDN suffix related code from `azureconst.go`
* Fix SwarmMode China deployment failure bug by extracting docker engine download repo and docker compose download URL.
* Update testcases.

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/499)

Fixes #1066

### CODE REVIEW

**Comment by wangtt03** [comment]
@colemickens Could you please review this PR? Thanks!

**Comment by raulfiru** [comment]
I'm back in the office only on Tuesday - at the moment i have crappy internet so the latest on Tuesday morning (monday evening for you) i'll regen and deploy

**Comment by raulfiru** [comment]
might me wrong here but i guess you are making that pull request from a private repo so i can't get the branch... and i can't merge into a new brach here since i don't have access (obviously)... So i'm stuck - can get the changes locally.

```
git remote add --noes not work
git fetch origin pull/ID/head:BRANCHNAME --does not work either
```

another question: do you use the location to determine the right FQDN or have another attribute?

this is what i set:
```
"location": "germanycentral",
```

**Comment by colemickens** [comment]
The repo is open, something like this should work:

```
git remote add ccg git@github.com:/ChinaCloudGroup/acs-engine.git
git remote update
git checkout sovereign_cloud
git reset --hard ccg/sovereign_cloud
```

**Comment by raulfiru** [comment]
worked! 

didn't have to change vm image name and now works like a charm... i still ssh-ed on the master (11PM... will try remote tomorrow)

```
azureuser@k8s-master-12051246-0:~$ kubectl get services
NAME         CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   10.0.0.1     <none>        443/TCP   7m
azureuser@k8s-master-12051246-0:~$ kubectl get nodes
NAME                        STATUS                     AGE
k8s-agentpool1-12051246-0   Ready                      7m
k8s-agentpool1-12051246-1   Ready                      7m
k8s-agentpool1-12051246-2   Ready                      7m
k8s-master-12051246-0       Ready,SchedulingDisabled   7m
```

great - tnx guys! you can merge from my point of view ;)

PS: didn't do any regression for non-Azure.de

**Comment by wangtt03** [comment]
@raulfiru I am so happy to hear that! Thanks! 👍

**Comment by benkoller** [comment]
I had to change the osImageVersion as `16.04.201703070` seems not to be available in germanycentral (?):

```
~ ❯❯❯ az vm image list --all --location germanycentral | jq  '[ .[] | select( .sku | contains ("16.04")) ]'
[
  {
    "offer": "UbuntuServer",
    "publisher": "Canonical",
    "sku": "16.04-LTS",
    "urn": "Canonical:UbuntuServer:16.04-LTS:16.04.201701130",
    "version": "16.04.201701130"
  },
  {
    "offer": "UbuntuServer",
    "publisher": "Canonical",
    "sku": "16.04.0-LTS",
    "urn": "Canonical:UbuntuServer:16.04.0-LTS:16.04.201604203",
    "version": "16.04.201604203"
  },
  {
    "offer": "UbuntuServer",
    "publisher": "Canonical",
    "sku": "16.04.0-LTS",
    "urn": "Canonical:UbuntuServer:16.04.0-LTS:16.04.201605161",
    "version": "16.04.201605161"
  },
  {
    "offer": "UbuntuServer",
    "publisher": "Canonical",
    "sku": "16.04.0-LTS",
    "urn": "Canonical:UbuntuServer:16.04.0-LTS:16.04.201609071",
    "version": "16.04.201609071"
  },
  {
    "offer": "UbuntuServer",
    "publisher": "Canonical",
    "sku": "16.04.0-LTS",
    "urn": "Canonical:UbuntuServer:16.04.0-LTS:16.04.201610200",
    "version": "16.04.201610200"
  }
]
```

After correcting the image version the master vm extension deployment fails as the custom script exits with `1`.

Checking `/var/log/azure/cluster-provision.log` I get a lot of these:

```
+ /usr/local/bin/kubectl cluster-info
The connection to the server localhost:8080 was refused - did you specify the right host or port?
+ '[' 1 = 0 ']'
+ sleep 1
+ for i in '{1..600}'
+ '[' -e /usr/local/bin/kubectl ']'
```

Any hints where I should look to troubleshoot further?

**Comment by wangtt03** [comment]
It seems that the apiserver is not started correctly. Please check if hyperkube docker image is pulled, could you please attach the docker ps output?

**Comment by benkoller** [comment]
I suspect I mangled the location configuration. Which points are relevant for setting the target environment?

**Comment by benkoller** [comment]
So, now I manually adjusted the `targetEnvironment` and `fqdnEndpointSuffix` in `azuredeploy.parameters.json` to get this to work as I still lack the understanding how and where you derive `containerService.Location`. I really would like to jump a bit deeper in this, but as of right now I lack the time to do so - sorry about that.

How did you plan on setting the location during the template generation?

**Comment by raulfiru** [comment]
@benkoller it worked like a charm for me... did you get the pull request as @colemickens wrote above (i had to change some paths) or you used the master branch?

You seam to know what you are doing (much more then me to be honest), but what i did is to get the pull locally (overwrite master), then follow the install guide where i've overwritten:
```
%GOPATH%\src\github.com\Azure\acs-engine
```
with the git folder that has the pull

then followed all the other steps and worked...

**Comment by benkoller** [comment]
I am on the `sovereign_cloud`-Branch and followed the above instructions to the letter. In the end all I had to do was the manual change of both FQDN and TargetEnv after template generation, but it doesn't seem like its the method intended by @wangtt03 

I unfortunately can't deploy the first workloads today but I'll deploy some stuff tomorrow. Until now all seems fine though.

**Comment by raulfiru** [comment]
that is strange indeed... i tried on 2 azure germany subscriptions.

on one worked every time (tried 3 times) - cluster-provision.log:

```
+ for i in '{1..600}'
+ '[' -e /usr/local/bin/kubectl ']'
+ /usr/local/bin/kubectl cluster-info
The connection to the server localhost:8080 was refused - did you specify the right host or port?
+ '[' 1 = 0 ']'
+ sleep 1
+ for i in '{1..600}'
+ '[' -e /usr/local/bin/kubectl ']'
+ /usr/local/bin/kubectl cluster-info
Kubernetes master is running at http://localhost:8080
```

on another..also tried 3 times... side by side with different dns names and then i reused the exact same file and deleted the one in the other subscription (i thought it could be the names/certs)

```
+ sleep 1
+ for i in '{1..600}'
+ '[' -e /usr/local/bin/kubectl ']'
+ /usr/local/bin/kubectl cluster-info
The connection to the server localhost:8080 was refused - did you specify the right host or port?
+ '[' 1 = 0 ']'
```

failed every time... same as @benkoller

**Comment by colemickens** [comment]
It's almost surely because of ServicePrincipal problems. I'm guessing the SP that you are using only has permissions on one of the subscriptions.

**Comment by raulfiru** [comment]
i'm using different SP for each.

and everything else gets provisioned except: k8s-master-12051246-0/cse0 - and that looks like a script that run on the vm:
```
"commandToExecute": "[concat('/usr/bin/nohup /bin/bash -c \"/bin/bash /opt/azure/containers/provision.sh ',variables('tenantID'),' ',variables('subscriptionId'),' ',variables('resourceGroup'),' ',variables('location'),' ',variables('subnetName'),' ',variables('nsgName'),' ',variables('virtualNetworkName'),' ',variables('routeTableName'),' ',variables('primaryAvailablitySetName'),' ',variables('servicePrincipalClientId'),' ',variables('servicePrincipalClientSecret'),' ',variables('clientPrivateKey'),' ',variables('targetEnvironment'),' ',variables('networkPolicy'),' ',variables('fqdnEndpointSuffix'), ' >> /var/log/azure/cluster-provision.log 2>&1 &\" &')]"
```

they run fine on agents but not on the master. It hangs there until the loop (600 sec) timesout

**Comment by colemickens** [comment]
@raulfiru Yes, it will fail if the SP is invalid. Please follow the troubleshooting steps here to rule out the SP credentials being invalid or having wrong permissions: https://github.com/Azure/acs-engine/blob/master/docs/kubernetes.md#misconfigured-service-principal

**Comment by raulfiru** [comment]
you were right
```
Apr 18 23:42:39 k8s-master-12051246-0 docker[6315]: E0418 23:42:39.796008    6659 kubelet_node_status.go:70] Unable to construct api.Node object for kubelet: failed to get external ID from cloud provider: autorest#WithErrorUnlessStatusCode: POST https://login.microsoftonline.de/b1904b97-e54a-4cb1-8031-6e578f3ad607/oauth2/token?api-version=1.0 failed with 400 Bad Request: StatusCode=400
```

i'll check the SP tomorrow and retry. 

tnx and good night for now

**Comment by raulfiru** [comment]
works! 

It was my mistake as i was using in the other case the SP name and not the ID.

now, both subscriptions work

**Comment by colemickens** [comment]
@raulfiru and to confirm, you didn't have to edit anything for FQDN or ubuntu image? I'd like to understand why this isn't working for @benkoller, if that is indeed still the case.

**Comment by raulfiru** [comment]
regarding the azure VM image - while the default image in the azuredeply.json is 16.04.201703070, the one is azuredeply.parameters.json is 16.04.201701130 - so it worked and i didn't change anything.

**Comment by benkoller** [comment]
@colemickens I have to power down the current deployment and will be out of the office until Monday so I can't check again. Until then I'd like to rule out I missed something during template generation, there are no additional steps to perform apart from `./acs-engine examples/kubernetes.json`, correct? Do I need to add a `"location": "AzureGermanCloud"` or similar to the `kubernetes.json`?

**Comment by wangtt03** [comment]
There should be a location field target to 'germanycentral' or 'germanynortheast'

**Comment by benkoller** [comment]
Can you share / add an updated `example/Kubernetes.json`?

**Comment by wangtt03** [comment]
{
  "apiVersion": "vlabs",
  "location":"germanycentral",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes"
    },
    "masterProfile": {
      "count": 1,
      "dnsPrefix": "germany-m01",
      "vmSize": "Standard_D2_v2"
    },
    "agentPoolProfiles": [
      {
        "name": "agentpool1",
        "count": 3,
        "vmSize": "Standard_D2_v2",
        "availabilityProfile": "AvailabilitySet"
      }
    ],
    "linuxProfile": {
      "adminUsername": "azureuser",
      "ssh": {
        "publicKeys": [
          {
            "keyData": "<your_ssh_public_key>"
          }
        ]
      }
    },
    "servicePrincipalProfile": {
      "servicePrincipalClientID": "<your_client_id>",
      "servicePrincipalClientSecret": "<your_client secret>"
    }
  }
}

Hope it helps!😊

**Comment by benkoller** [comment]
Thx @wangtt03, deployment went without a hitch with the location field. I had used `AzureGermanCloud` instead of `germanycentral`, thx for your example.

**Comment by dmitsh** [comment]
Did it work with "germanycentral" ?

**Comment by benkoller** [comment]
@dmitsh yes, sorry for being unclear about it. I had used `"location": "AzureGermanCloud"`, which did **not** work, but when I started using `"location":"germanycentral"` the template generation started working flawlessly. Here is the (redacted) kubernetes.json I used as input:

```
{
  "apiVersion": "vlabs",
  "location": "germanycentral",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes"
    },
    "masterProfile": {
      "count": 1,
      "dnsPrefix": "...",
      "vmSize": "Standard_D2_v2"
    },
    "agentPoolProfiles": [
      {
        "name": "agentpool1",
        "count": 3,
        "vmSize": "Standard_D2_v2",
        "availabilityProfile": "AvailabilitySet"
      }
    ],
    "linuxProfile": {
      "adminUsername": "...",
      "ssh": {
        "publicKeys": [
          {
            "keyData": "ssh-rsa ..."
          }
        ]
      }
    },
    "servicePrincipalProfile": {
      "servicePrincipalClientID": "...",
      "servicePrincipalClientSecret": "..."
    }
  }
}
```

**Comment by raulfiru** [comment]
@wangtt03 @colemickens do you need any more support to merge this pull - except with the conflicts where i don't think we can help.

**Comment by wangtt03** [comment]
@benkoller, the location field should be `germanycentral` or `germanynortheast`, once the ARM template is generated, the `targetEnvironment` will be automatically set to `AzureGermanCloud`. Glad to know it works:)

**Comment by benkoller** [comment]
@wangtt03 any plans of merging this into master and adding the related documentation?

**Comment by wangtt03** [comment]
@colemickens Could you please review this PR? Thanks in advance!

**Comment by wangtt03** [comment]
@anhowe Could you please also review this PR? Thanks!

Changes:
* extract all the hard coded download urls into central place.
    * kubernetes: networkpolicy download urls, hard code master fqdn suffix in shell script.
    * swarm(mode): docker engine download urls and docker-compose download url.
* update `installDocker` function in swarm cluster shell script, use `apt-get` to install.
* make osimage configurable to meet the different cloud. (ubuntu image version is different in germany)
* add hard coded `germany` and `usgov` locations in azureconst.go/Get-AzureConstants.ps1

**Comment by anhowe** [comment]
@wangtt03 - please remove all the *expected* files, and resend

**Comment by wangtt03** [comment]
@anhowe Done!

**Comment by shrutir25** [comment]
@anhowe how do we proceed with this PR ?

**Comment by shrutir25** [comment]
@JackQuincy - could you review this PR ?

**Comment by DonMartin76** [comment]
What's the status of this? Any chance this will make it into `master` anytime soon? It would be really awesome if that is the case. We have other teams also targeting Kubernetes, but who need the German Cloud.

**Comment by shrutir25** [comment]
@anhowe how do we proceed with this PR ?

**Comment by raulfiru** [comment]
i would really appreciate it if we can merge this - and we don't have to jump though hoops in order to deploy k8s

**Comment by seanknox** [comment]
Hey folks, sorry this has sat so long. I've made it a priority to be merged in and available for acs-engine v0.5.0.

**Comment by seanknox** [comment]
@wangtt03 can you rebase and we'll run in CI?

**Comment by wangtt03** [comment]
@seanknox Thanks for reviewing! Please see my updates:)

**Comment by seanknox** [comment]
Thanks @wangtt03, will plan to merge tomorrow after others have had a chance to comment.

cc @Azure/acs-engine-core

**Comment by JackQuincy** [comment]
What are thoughts around taking this before we can get CI verifying it is working properly? As @amanohar brought up?

**Comment by seanknox** [comment]
@JackQuincy as acs-engine is OSS (not supported), we could merge this in now and then add regression test for those regions. I don't see @amanohar original comments...

**Comment by JackQuincy** [comment]
my bad it was @weinong  he said

> To test it, you need different Azure subscriptions for respective sovereign cloud. the current CI won't be effective.

**Comment by raulfiru** [comment]
if you need to make a quick test in Azure Germany after it was merged - let me know.

**Comment by dmitsh** [comment]
A quick question. Would Azure Public CDN be available from within BlackForest? Could someone verify this link https://acs-mirror.azureedge.net/cni/cni-amd64-latest.tgz

**Comment by DonMartin76** [comment]
On vacation, cannot assist currently. Sorry!

**Comment by raulfiru** [comment]
it works!

**Comment by raulfiru** [comment]
we are actually waiting for this merge - and not deploying k8s at the moment(did it once manually..but we would not like to create scripts based on a pull request). could you give us a feeling about this? will it take a few days? weeks?

**Comment by jackfrancis** [comment]
@raulfiru This needs to be rebased, after which with another sanity-check review we can merge if existing E2E tests pass. As has been mentioned above, once everything passes and merges, we can add coverage for these two clouds to make them first class E2E citizens going forward.

**Comment by optiomas** [comment]
+1

**Comment by jackfrancis** [comment]
@wangtt03 could you rebase this on master, or give me admin perms on ChinaCloudGroup:sovereign_cloud and I'd be happy to do it myself. Thanks either way!

**Comment by wangtt03** [comment]
@jackfrancis ,I will rebase this PR, and I will give you the admin permission also.

**Comment by jackfrancis** [comment]
@wangtt03 I checked out the `ChinaCloudGroup` fork of `acs-engine`, but don't have permissions to push (I have some minor changes that fix `make test-style` warnings).

**Comment by jackfrancis** [comment]
@wangtt03 false alarm, I hadn't accepted the collaboration invite, all works now! thanks!

File: parts/configure-swarm-cluster.sh [hunk 1]
```diff
@@ -161,16 +163,13 @@ echo "Installing and configuring docker"
...
-    wget --tries 4 --retry-connrefused --waitretry=15 -qO- https://get.docker.com | sh
-    if [ $? -eq 0 ]
-    then
-      # hostname has been found continue
-      echo "Docker installed successfully"
-      break
-    fi
-    sleep 10
-  done
+  apt-get update
```
Comment (anhowe) [inline parts/configure-swarm-cluster.sh]: all "apt" commands should have retries around them.  Please see other areas of code where we do this.

File: parts/configure-swarm-cluster.sh [hunk 2]
```diff
@@ -159,18 +161,18 @@ echo "$HOSTADDR $VMNAME" | sudo tee -a /etc/hosts
 
 echo "Installing and configuring docker"
 
+# simple general command retry function
+retrycmd_if_failure() { for i in 1 2 3 4 5; do $@; [ $? -eq 0  ] && break || sleep 5; done ; }
```
Comment (jackfrancis) [inline parts/configure-swarm-cluster.sh]: @anhowe See this function and the below invocations to address your retry apt commands suggestion.

**Comment by optiomas** [comment]
🚀

**Comment by raulfiru** [comment]
checking it today/tomorrow awesome!

**Pull request merged**